### PR TITLE
Improvements to Query Result serialization code

### PR DESCRIPTION
**Pull request by arikfr**
## What type of PR is this? (check all applicable)

- [x] Bug Fix
- [x] Other

## Description

1. Make the CSV serialization of dates safer by handling some edge cases (like invalid or empty values).
2. Improve performance (x10 in my benchmark) of serialization by switching from using [`dateutils.parser.parse`](https://dateutil.readthedocs.io/en/stable/parser.html#dateutil.parser.parse) to [`dateutils.parser.isoparse`](https://dateutil.readthedocs.io/en/stable/parser.html#dateutil.parser.isoparse).

Benchmarking was done with the following script:

```python
from redash import create_app
from redash.models import QueryResult
from redash.utils import json_dumps
from redash.serializers import serialize_query_result_to_csv
import pstats
import cProfile


data = {
    "rows": [
        {"datetime": "2019-05-26T12:39:23.026Z", "date": "2019-05-26"},
    ], 
    "columns": [
        {"friendly_name": "date", "type": "datetime", "name": "datetime"},
        {"friendly_name": "date", "type": "date", "name": "date"}
    ]
}

for i in xrange(5000):
    data['rows'].append({"datetime": "2019-05-26T12:39:23.026Z", "date": "2019-05-26"})


def bench():
    app = create_app()
    app.config['TESTING'] = True
    app.config['SERVER_NAME'] = 'localhost'
    app_ctx = app.app_context()
    app_ctx.push()

    with app.test_request_context('/'):
        query_result = QueryResult(data=json_dumps(data))
        # serialize_query_result_to_csv(query_result)
        cProfile.runctx('serialize_query_result_to_csv(query_result)', globals(), locals(), filename="profile")
        p = pstats.Stats('profile')
        p.sort_stats('cumulative').print_stats(10)


if __name__ == '__main__':
    bench()
```

If I knew the issue would be in `_convert_datetime`, I could make a simpler benchmarking script, but 🤷‍♂ .

The results before the change were:

```
         3782685 function calls (3721728 primitive calls) in 58.708 seconds
```

And after the change:

```
         306441 function calls (305520 primitive calls) in 4.960 seconds
```

### CODE REVIEW

File: redash/serializers/query_result.py [hunk 1]
```diff
@@ -19,42 +20,39 @@ def _convert_bool(value):
...
-    if not value:
-        return value
-
-    parsed = parse_date(value)
-
-    return parsed.strftime(_convert_format(current_org.get_setting('date_format')))
 
-
-def _convert_datetime(value):
+def _convert_datetime(value, fmt):
```
Comment (rauchy) [inline redash/serializers/query_result.py]: Can't this all be summed up to

```suggestion
def _convert_datetime(value, fmt):
    try:
        parsed = parse_date(value)
        return parsed.strftime(fmt)
    except Exception:
        return value       
```

File: tests/serializers/test_query_results.py [hunk 2]
```diff
@@ -44,6 +46,15 @@ def test_serializes_datatime_with_correct_format(self):
         self.assertEqual(rows[0]['datetime'], '26/05/19 12:39')
         self.assertEqual(rows[1]['datetime'], '')
         self.assertEqual(rows[2]['datetime'], '')
+        self.assertEqual(rows[2]['date'], '')
```
Comment (rauchy) [inline tests/serializers/test_query_results.py]: Duplicate?

File: redash/serializers/query_result.py [hunk 3]
```diff
@@ -19,42 +20,39 @@ def _convert_bool(value):
...
-    TYPE_DATE: _convert_date,
-    TYPE_DATETIME: _convert_datetime
-}
+def _get_column_lists(columns):
+    date_format = _convert_format(current_org.get_setting('date_format'))
+    datetime_format = _convert_format('{} {}'.format(current_org.get_setting('date_format'), current_org.get_setting('time_format')))
 
+    special_types = {
+        TYPE_BOOLEAN: _convert_bool,
+        TYPE_DATE: rpartial(_convert_datetime, date_format),
```
Comment (rauchy) [inline redash/serializers/query_result.py]: 👍

File: tests/serializers/test_query_results.py (hunk 2 above)
Comment (arikfr) [reply tests/serializers/test_query_results.py]: Why?

Comment (rauchy) [reply tests/serializers/test_query_results.py]: With row 52.

Comment (arikfr) [reply tests/serializers/test_query_results.py]: Yep.

**Pull request merged**